### ISSUE TITLE
non-shuffling data loaders can affect random states, thus the results of shuffling data loaders.

### ISSUE THREAD

**Issue by zym1010** [pytorch/pytorch#11062]
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description

See the title. While it's arguable this behavior is bug or not, I feel it's not very intuitive. Maybe it's good to doc this.

For an actual example, see below.

In the example, I have two data loaders, one shuffling for training, one non-shuffling for validation. Turns out, the order of data sample points for the shuffling training loader depends on how many times I call the non-shuffling validation loader.

## Code example

```python
# In[1]:


from torch.utils.data import DataLoader
from torch.utils.data.dataset import TensorDataset
import torch
print(torch.__version__)


# In[2]:


import numpy as np
np.random.seed(0)

# In[3]:


dataset_train = np.random.randint(100, size=(10,)).astype(np.float32)
dataset_val = np.random.randint(100, size=(10,)).astype(np.float32)
print('train\n',dataset_train)
print('val\n',dataset_val)

# train
# [44. 47. 64. 67. 67.  9. 83. 21. 36. 87.]
# val
# [70. 88. 88. 12. 58. 65. 39. 87. 46. 88.]

# In[4]:


dataloader_train = DataLoader(TensorDataset(torch.tensor(dataset_train)),
                                            batch_size=10, shuffle=True)
dataloader_val = DataLoader(TensorDataset(torch.tensor(dataset_val)),
                         batch_size=10, shuffle=False)
def loop_one(dataloader_this):
    for (x,) in dataloader_this:
        print(x)




# In[5]:


# first time, only train
np.random.seed(0)
torch.manual_seed(0)
loop_one(dataloader_train)

# no val, we get the following for dataloader_train
# tensor([67., 21.,  9., 64., 44., 36., 47., 83., 87., 67.])

# In[6]:


# second time, val + train
np.random.seed(0)
torch.manual_seed(0)
loop_one(dataloader_val)
loop_one(dataloader_train)

# one val
# tensor([67., 21.,  9., 83., 64., 36., 47., 87., 67., 44.])

# In[7]:


# second time, val + val + train
np.random.seed(0)
torch.manual_seed(0)
loop_one(dataloader_val)
loop_one(dataloader_val)
loop_one(dataloader_train)

# two val
# tensor([21., 64., 67., 36., 87., 83.,  9., 44., 47., 67.])

# In[8]:


# second time, val + val + train
np.random.seed(0)
torch.manual_seed(0)
loop_one(dataloader_val)
loop_one(dataloader_val)
np.random.seed(0)
torch.manual_seed(0)
loop_one(dataloader_train)

# reseed
# tensor([67., 21.,  9., 64., 44., 36., 47., 83., 87., 67.])
```

## System Info

I ran this under a Singularity container with Ubuntu 14.04 host, and I got the following system info by running that `collect_env.py` script in the container.

```
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX TITAN Black
GPU 1: Tesla K40c

Nvidia driver version: 384.111
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

**Comment by fmassa** [comment]
This happens because we unconditionally [generate a random number in the `DataLoaderIter`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L255), which changes the state of the rng.
This is kind of documented in [this part](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L433-L441).

I think that we could reset the state of the rng to the state it was before the generation of this number so that this doesn't happen.

**Comment by zym1010** [comment]
Thanks. I agree with your fix to this problem (I skimmed over the code for data loader a while ago). As for the doc, I don’t think reading the doc in its current state would make people realize the issue I describe here.

To express my concern in another way, I think making behaviorally deterministic operations affect random states is certainly unintuitive; for example, you would not expect that running a few matrix multiplications in numpy can affect random number generation that follows those matrix multiplications.

In addition, I think it would be nice to make data loaders have their own random states without modifying the system ones. 

Yimeng Zhang

**Comment by soumith** [comment]
> I think making behaviorally deterministic operations affect random states is certainly unintuitive; for example, you would not expect that running a few matrix multiplications in numpy can affect random number generation that follows those matrix multiplications.

This is absolutely the right experience to aim for. We dont want behaviorally (or perceptually) deterministic operations to affect random state.

One easy fix is to (as you suggested), cache a one-time base_seed in the DataLoader itself, and `DataLoaderIter`'s base seed is deterministically incremented starting from DataLoader's base seed.
